nncf/common/quantization/quantizer_propagation/structs.py

```diff
@@ -0,0 +1,91 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from enum import Enum
+from typing import List
+from typing import Optional
+from typing import Set
+
+from nncf.common.quantization.structs import QuantizerConfig
+from nncf.common.quantization.structs import UnifiedScaleType
+
+
+class QuantizationTrait(Enum):
+    """
+    General, hardware-agnostic specifications for the relation of operators to quantization.
+    Hardware-specific quantization configuration is handled elsewhere.
+    """
+    NON_QUANTIZABLE = -1
+    QUANTIZATION_AGNOSTIC = 0
+    INPUTS_QUANTIZABLE = 1
+
+    # For embeddings, the inputs are integer and are used to index into the weight of the layer,
+    # therefore if the weight is already quantized there may be no need to quantize the outgoing tensor.
+    # TODO: unify scales for such operations if they become linked through a downstream op (such as
+    # two embeddings added to each other)
+    OUTPUT_QUANTIZATION_AS_WEIGHTS = 2
+
+    # A concat node should ultimately either have all of its inputs quantized with the same configuration
+    # or none at all. This cannot be determined ahead of time, hence the special trait.
+    CONCAT = 100
+
+
+class PropagatingQuantizer:
+    """
+    Used in conjunction with QuantizerPropagationStateGraph to keep track of
+    the allowed quantization configs corresponding to the model operation node
+    whose inputs it quantizes, and also of the nodes/edges in the model control
+    graph that this quantizer affects. It should be moved against the data flow of
+    the model, tracking the affected nodes and edges of
+    QuantizerPropagationStateGraph. No actual quantization modules are used here,
+    only the associated configs (such as bitwidths, modes, signed/unsigned
+    attributes etc.)
+    """
+
+    def __init__(self, id_: int, quant_configs: List[QuantizerConfig], init_location_node_key: str,
```

Comment:
Could you please add a docstring for this method?

Reply:
Done